examples/traffic_histories_to_observations.py

```diff
@@ -1,130 +1,302 @@
+                with open(outfile, "wb") as of:
+                    pickle.dump(data, of)
 
-            for veh_id in history_vehicles:
-                try:
-                    smarts.attach_sensors_to_vehicles(agent_spec.interface, {veh_id})
-                except ControllerOutOfLaneException:
-                    logger.warning(f"{veh_id} out of lane, skipped attaching sensors")
-                    vehicles_off_road.add(veh_id)
+        self._smarts.destroy()
 
-            valid_vehicles = {v for v in history_vehicles if v not in vehicles_off_road}
-            obs, _, _, dones = smarts.observe_from(list(valid_vehicles))
-            _record_data(smarts.elapsed_sim_time, obs, collected_data)
+    def _record_data(
+        self,
+        collected_data,
+        current_vehicles,
+        off_road_vehicles,
+        selected_vehicles,
+    ):
+        # In future, allow for explicit mapping of vehicle_ids to agent_ids.
+        agent_ids = cycle(self._env.agent_specs.keys())
 
-        # an example of how we might save the data per car
-        observation_folder = "collected_observations"
-        if not os.path.exists(observation_folder):
-            os.makedirs(observation_folder)
-        for car, data in collected_data.items():
-            outfile = f"{observation_folder}/{scenario.name}_{scenario.traffic_history.name}_{car}.pkl"
-            with open(outfile, "wb") as of:
-                pickle.dump(data, of)
+        # Attach sensors to each vehicle
+        valid_vehicles = (current_vehicles - off_road_vehicles) & selected_vehicles
+        for veh_id in valid_vehicles:
+            agent_id = next(agent_ids)
+            agent_interface = self._env.agent_specs[agent_id].interface
+            try:
+                self._smarts.attach_sensors_to_vehicles(agent_interface, {veh_id})
+            except ControllerOutOfLaneException:
+                self._logger.warning(f"{veh_id} out of lane, skipped attaching sensors")
+                off_road_vehicles.add(veh_id)
 
-    smarts.destroy()
+        # Get observations from each vehicle and record them
+        obs = dict()
+        obs, _, _, _ = self._smarts.observe_from(list(valid_vehicles))
+        resolutions = {}
+        for id_ in list(obs):
+            if obs[id_].top_down_rgb:
+                resolutions[id_] = obs[id_].top_down_rgb.metadata.resolution
+            ego_state = obs[id_].ego_vehicle_state
+            if ego_state.lane_index is None:
+                del obs[id_]
+            else:
+                mission = self._missions[ego_state.id]
+                if mission:
+                    # doh! ego_state is immutable!
+                    new_ego_state = ego_state._replace(mission=mission)
+                    obs[id_] = replace(obs[id_], ego_vehicle_state=new_ego_state)
+        # TODO: handle case where neighboring vehicle has lane_index of None too
+        obs = self._std_obs_wrapper.observation(obs)
```

Comment:
- Why did we not just return the `smarts.sensors.Observation` directly?
- Why did we use the `FormatObs` wrapper here to reformat the observation to `StdObs`?
- By using `FormatObs` wrapper here, we would be making this standard format mandatory for all users. Some users may not realise that this wrapper has been applied internally.
- Can we simply return `smarts.sensors.Observation`?

Reply:
I did it this way b/c this is what @saulfield did originally in PR #1461.  So I'll let him answer these questions.